stan/math/rev/fun/generalized_inverse.hpp

```diff
@@ -21,17 +22,19 @@ namespace internal {
 template <typename T1, typename T2>
 inline auto generalized_inverse_lambda(T1& G_arena, T2& inv_G) {
   return [G_arena, inv_G]() mutable {
+    auto ones1
+        = Eigen::VectorXd::Constant(std::min(G_arena.rows(), inv_G.cols()), 1);
+    auto ones2
+        = Eigen::VectorXd::Constant(std::min(inv_G.rows(), G_arena.cols()), 1);
     G_arena.adj()
         += -(inv_G.val_op().transpose() * inv_G.adj_op()
              * inv_G.val_op().transpose())
-           + (-G_arena.val_op() * inv_G.val_op()
-              + Eigen::MatrixXd::Identity(G_arena.rows(), inv_G.cols()))
+           + add_diag(-G_arena.val_op() * inv_G.val_op(), ones1)
```

Comment:
Why did you switch this to `add_diag()`?

Reply:
So that 2 identity matrices aren't created and the matrix ops only have to add to the diagonal. I am just learning Eigen so I'm not sure if they've optimized adding identity matrices or how creating an identity matrix affects memory.

Reply:
Eigen is clever enough to know that when you give it an `Eigen::MatrixXd::Identity` it will most likely just use a single scalar to denote the whole matrix (I can't remember what exactly they do but they are clever about it)

Reply:
Should I change it back to identity? I noticed some speedup in the tests but that could just be an artifact

Reply:
I would think that's an artifact I think you can change it back